### ISSUE TITLE
Combined query to get sermons and teaching series

### ISSUE THREAD

**Issue by wesgate** [CrossingsCommunityChurch/RockTheme#44]
- Needs to combine sermons and teaching series, order by date (newest to latest)
- Needs to be queryable by Speaker, Book of the Bible, or Topic

**Comment by camrun91** [comment]
We should be able to use this and get good performance by using Persisted datasets. https://community.rockrms.com/documentation/bookcontent/14/185#persisteddatasets 
I have added this page under CMS Configuration and we should be able to set up any datasets we want/need.

**Comment by wesgate** [comment]
@racingcow 

Hey man, 

We have a page /media that is supposed to contain all messages and all teaching series content channel items (these content channel items are setup in the same way, but they needed to be broken out for the app).

With the current setup using the out of the box block from Rock (Content Channel View), we can only have a single content channel item assigned (currently we have sermon messages), so we need to write a query that will get all messages and teaching messages, merge them and sort them by active date. Once we have the data in the way we need it, we can then write our own shortcode to handle displaying them

I think ideally we would setup a query that we would store as a persisted dataset (pretty sure we can do this with just some lava and get pretty close and just store the result as JSON in the persisted dataset)

If you wanted to take a look at what newspring I belive is doing:


Wes

**Comment by wesgate** [comment]
@racingcow 

As an example of a persisted data set, if you get the latest from master, I have a new file in the Data folder - latestMessage.lava.  I turned your latest message query into a persisted dataset and then I am using it on the /Sections/latestMessage.lava file (we could swap out button to use this persisted set as well)

You can get to the PersistedDatasets from CMS Configuration->Persistent datasets (FYI - i had to do a clean build before the icon showed up for me)

You can see where/how I setup the persisted dataset there.

FYI - on the includes in the dataset, i had to have to look outside the Theme (because the backend uses the 'Rock' theme and we need it to look in 'RockTheme'

Looks like this:

{% include '~/Themes/RockTheme/Assets/Lava/crossings/Data/latestMessage.lava' %}

I am sure that would also solve the issues with that location page not finding the lava file

Wes

@camrun91

**Comment by camrun91** [comment]
@racingcow @wesgate I have a simple dataset that appears to be working in test that is pulling the media items for teaching and sermon media. I will put it in a shortcode and stuff and let you all know where it is.

**Comment by racingcow** [comment]
That looks good to me, @camrun91. I added a commit to look up the ids of the content channels by name in case it helps with dev vs prod later.

@wesgate Are there any other things that I could help with?

**Comment by wesgate** [comment]
@racingcow 

Yeah, I think the additional steps we need to take now that we have the dataset is figuring out a way to paginate the results and also query the dataset by Tags and by Speaker (you will see the filters above the media results).

The filter drop-downs are currently just hard-coded. We need to probably dynamically create those filters based on the tags in the tag categories. We currently have a category of tags for "Topics" (we need to make another set that is just books of the bible) and then we would need to setup the list of speakers as well and include them.

Once we are generating those, we need to make it so that when you select a topic, speaker or book of the bible, it will query the dataset and filter the results. I get how we did this with entities (simply a where clause), but I am not super sure how we do this with lava when using the persisted dataset . (other than looping through the entire dataset and running a bunch of checks to see if a message contains the selected tag or speaker). We probably should make it where folks could maybe even combine the queries (select a book of the bible or topic along with a speaker)

Let me know if you have any questions on that

Wes

Wes

**Comment by camrun91** [comment]
@wesgate @racingcow I will work today on cleaning the dataset some and getting it to have the info we need to have to display it on the page. 

I had also want to get a list going of things we think would benefit from being made to use datasets.

**Comment by racingcow** [comment]
@wesgate ,

I can think of a way (I think) to manually filter and page the persisted dataset with looping, but not sure that there is a better way, short of using their API or doing something super custom or un-"rock like", provided I'm making good assumptions about what you want.

My guess is that we'd like to find a way to make it so the page doesn't reload every time someone changes a filter or page? Or are you thinking it is okay to have reloads?

The "Content Channel View" (`\RockWeb\Blocks\Cms\ContentChannelView.ascx.cs`) is sort of what we want (caching, filtering via dynamic params, paging), but it only works on a single channel and uses full page refreshes every time you switch pages. 

Also, I'm not yet sure if a persisted dataset can have more than 1000 records. It might, but I'm not sure yet... maybe it can, and I wouldn't think it would be a problem for a while, but let me know if you think they'd go over 1000 combined items.

**Comment by wesgate** [comment]
@racingcow - the manual way was the only way I could think of off the top of my head as well since we don have the usual 'entity' filters that we can use on the persisted dataset (at least I don't think we can use a 'Where' on the dataset)

I know from a UI perspective, it isn't super fun to make the page refresh every time we load a filter or go to the next page, but I would almost say we start there and see if we can get it working and then if we need to do some ajax or something and hit the API, or figure out a better way to retrieve the data set and filter it, we could do that. 

Wes

**Comment by camrun91** [comment]
Look at some of the Newspring stuff I think they are using the where filters on datasets.

**Comment by wesgate** [comment]
They at least look to be using limit and offset here on a for loop:

https://github.com/NewSpring/Rock/blob/master/RockWeb/Themes/NewSpring/assets/Lava/LAYOUT/card-list.lava


On Jul 9, 2020, at 11:48 PM, Cameron Llewellyn <notifications@github.com>
wrote:

﻿

Look at some of the Newspring stuff I think they are using the where
filters on datasets.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/CrossingsCommunityChurch/RockTheme/issues/44#issuecomment-656480183>,
or unsubscribe

**Comment by racingcow** [comment]
@wesgate ,

Sounds good. I added tags for the three categories you mentioned and a new shortcode to build the `li`s for each tag category on the media page. I could change it to 3 persisted datasets if you think the speed improvement would warrant it. Just let me know. I'll work on the filtering tomorrow unless you guys beat me to it.

![image](https://user-images.githubusercontent.com/1176637/87120875-894dbe80-c24f-11ea-8a75-c96cad639c18.png)

**Comment by wesgate** [comment]
@racingcow 

Looks good, can we make a persisted dataset of speakers that is generated from the 'Speaker' property from messages rather than keeping up with an additional tag list for that one? Essentially creating a dataset of unique Speakers found from the Message property of 'Speaker'. We are already selecting a Speaker that way (since that is a default option on a message and that is tied to a person in the CRM)

Wes

**Comment by camrun91** [comment]
So in the dataset of the message combined I have a speaker field that lists just the person's name. I wonder if we can create another dataset that has a value for the tags, speakers, and books of the bible that have been used as well as contains the dataset of the media items?

I dk if that makes sense or not.

**Comment by racingcow** [comment]
@camrun91 @wesgate 

I think it makes sense here to tie it all into a single persisted dataset. It doesn't seem likely that we'd need those things individually in other places. If we find that we do, maybe it would make sense to break it up into separate ones.

Tonight I worked on using the speakers from the content channel items instead of having to manage tags. This was more difficult than I thought. Since there will be more than one message with the same speaker, I needed to create a dynamic list using lava. At first, I tried using entity commands, but quickly discovered that there isn't a great way (or maybe any way) to create a distinct list from within a persisted data set in this way. I would need some sort of multi-valued thing to remember what has already gone into the list, and as far as I can tell, there is no way to create a collection, vector, or otherwise dynamic array. I tried to do it by building up a string and trying to match against it with a regex like this...

![image](https://user-images.githubusercontent.com/1176637/87218302-730f3380-c31f-11ea-884b-b8fb1f682e17.png)

That _kind of_ worked, but for whatever reason, it seems like you can't conditionally add the curly braces in JSON. An empty item appears in the JSON regardless of whether the `if match == false` line fires. Here is my test display code and output...

![image](https://user-images.githubusercontent.com/1176637/87218359-d8632480-c31f-11ea-931f-2d391b06417f.png)

![image](https://user-images.githubusercontent.com/1176637/87218360-df8a3280-c31f-11ea-9280-1bf350b4557c.png)

The blank value there was a second Marty Grubbs. It has been filtered out, but the empty item in the array still exists for reasons that I haven't chased down further. I could probably crack open the lava rendering engine and figure out why, but figured it was time to move on. Let me know if you see something that I've overlooked that could be causing it.

I switched to SQL, and that seems to work, but may give us compatibility issues in the future. I could build a View for this in the database if you think it would help. Here's the code...

![image](https://user-images.githubusercontent.com/1176637/87218431-97b7db00-c320-11ea-94f4-b904d6843909.png)

I'll move on to filtering the list and paging on the media page next. Let me know if you have any suggestions on this, though, and I can do those first.

**Comment by racingcow** [comment]
@wesgate  and @camrun91 ,

I added some commits with thoughts on one way we could do filtering. Have a look and, as always, let me know if you have any adjustments or other feedback.

**Comment by wesgate** [comment]
@racingcow commits look good to me man, the way you are filtering is about the best way I can think to do it.

We will eventually want to maybe consider combined queries as well (book={{book.Name}}&speaker={{speaker.Name}}? (so we may need to look to see if a param already exists and then create the url accordingly)

Using SQL in that way i think is just going to be a necessary evil to get things to work the way we want

**Comment by racingcow** [comment]
@wesgate @camrun91 ,

I think basic filtering is working now. I found a way to pull the speakers list without SQL, so I moved back to that. Also, it looked like entities could only be associated with one kind of tag category, so I added a new attribute for teaching and sermon messages called "Books". It's still backed by whatever tags are put in the "Bible Books Tags" category, though.

![image](https://user-images.githubusercontent.com/1176637/87384763-8f52e080-c56a-11ea-968f-be5832d0362f.png)

I will put a PR up for this. Let me know if you think there is more to do before we call this one done (more polish on the `/media` page, etc.)

**Comment by camrun91** [comment]
I think this one can be completed once it is working in Production. I think there will likely be enhancement requests made that would be associated with this but we will see.

**Comment by wesgate** [comment]
@racingcow any chance you can figure out a way to paginate these?

**Comment by wesgate** [comment]
@racingcow Also, can we populate the default text on the filters to whatever filter is active? So if it is on Marty Grubbs as speaker, set that filter to say Marty Grubs instead of Select a Speaker

**Comment by racingcow** [comment]
@wesgate ,

Sure. I'll get the rest of paging going and make sure that the selected items in the dropdowns are synched with the active filter values. I should also be able to allow multiple filters at the same time.

**Comment by racingcow** [comment]
@wesgate @camrun91 ,

I think paging and showing the selected filter items in the dropdowns are both working now. Let me know if you see anything else that should be changed.

**Comment by camrun91** [comment]
It looks to me that the next and previous buttons are showing but the number of items being shown are not being limited.

**Comment by wesgate** [comment]
Yeah, it's fine, I went in and fixed it, we just had to set a default for when the PageParameter wasn’t set

if you get the latest it will work

Wes

**Comment by wesgate** [comment]
@racingcow 

Hey David,

I did merge your changes and I set the default page parameter to 1 so that /media/ would just show the first 6. If there is a better way to do that, just change it. A couple more things:

1) Can we hide the back button or next button respectively if we are on the first or last page?

2) I didn't investigate as to why, but it seems that the Back button didn't work on page 2 (going back to page 1), it works on 3 or 4 going back though. Seems to be something with that first set

Thanks!

Wes

@camrun91
